### ISSUE TITLE
Environment name unique naming should also a function of time not just sha256 of environment.yaml

### ISSUE THREAD

**Issue by costrouc** [Quansight/conda-store#44]
# Current database model

Environment -> (specification_sha256, build_id)

build_id -> specification_sha256

specification_id (doesn't point to anything)

# Proposal 

If you build a conda environment at a set point in time it will result in a different set of packages than if it was built 1 day ago. Thus it doesn't make much sense for the artifact to have the specification sha 256 to be in the name.

I want to support a (rebuild) environment action. Thus I really don't think that a specficaition_id is useful when talking about an environment.

Environment -> build_id -> specification_sha256

**Comment by costrouc** [comment]
Thought a bit more about this issue. The core of the issue is that we need a reproducible way to specify a given environment.

Originally I made the environment name the `sha256(environment.yaml)` file which seemed to be a decent way to represent an environment e.g. `15813e6c72c2535da602b162b314e67a44efa4a1-envname`. However there is a major drawback of this choice a build of an environment at a different point in time will result in likely a different environment assuming that packages are not heavily pinned due to using a different redodata.json. Thus the unique representation of a given environment is a `function(sha256(environment.yaml), datetime)`. Sometimes it would be nice to periodically rebuild a given environment. Thus we need an easy way to represent a build. 

I propose `{sha256(environment.yaml)}-{math.floor(time.time())}-<environment-name>` as the build representation.

**Comment by costrouc** [comment]
This will not require database changes which would be nice!

**Comment by leej3** [comment]
Just a thought, ignore if not useful... i wonder whether sha256 of the output of `conda env export` is possible. A downside is that it requires a solved environment to be able to run the command. One nice detail is you won't need to care about the time of the env build. If you did want to support a faster, non-optimized, but faster solve, like mamba... rebuilds that do not guarantee the same solve would not be considered the same for these purposes.

**Comment by costrouc** [comment]
@leej3 thanks for that! So this really reminds me of a debate in the nix community. 

The idea is `sha256(package configuration, (sha256(dependency) for dependency in dependencies)) -> 15813e6c72c2535da602b162b314e67a44efa4a1-packagename`. This is the way that nixpkgs was for a long time and mostly is at this moment. And then as nixpkgs became more mature the builds themselves were reproducible (see https://r13y.com/). The base iso for nixos has 1660/1663 fully reproducible. Another interesting article https://www.tweag.io/blog/2020-09-10-nix-cas/. Nixpkgs is using the concept you have mentioned as `content addressable builds` e.g. if the build is reproducible we should with a sha256 of the build results be able to represent the package along with a strong guarantee about the compiled result which gives a lot of interesting properties.

To me `sha256(conda env export)` is about as close as conda can get to this content addressable concept. 

The major hurdle I see of this is that it is not possible to calculate `sha256(conda env export)` until the environment is build (due to pip). And in order to change the directory name you have to move conda the conda env (expensible in time) .... or use a symlink!

So I'd like to propose an idea that we have a symlink to that build that is the sha256(conda env export) -> 

`environment-name -> `{sha256(conda env export)}-name` -> `{sha256(environment.yaml)}-{math.floor(time.time()):016}-<environment-name>`

Since `{sha256(environment.yaml)}-{math.floor(time.time()):016}-<environment-name>` can be precomputed. 

I just ran `conda env export` and this was an example result I got 

```yaml
name: base
channels:
  - conda-forge
  - https://conda.anaconda.org/conda-forge
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=0_gnu
  - bzip2=1.0.8=h516909a_2
  - ca-certificates=2019.11.28=hecc5488_0
  - certifi=2019.11.28=py37_0
  - cffi=1.13.2=py37h8022711_0
  - chardet=3.0.4=py37_1003
  - conda=4.8.2=py37_0
  - conda-package-handling=1.6.0=py37h516909a_1
  - cryptography=2.8=py37h72c5cf5_1
  - idna=2.8=py37_1000
  - ld_impl_linux-64=2.33.1=h53a641e_8
  - libffi=3.2.1=he1b5a44_1006
  - libgcc-ng=9.2.0=h24d8f2e_2
  - libgomp=9.2.0=h24d8f2e_2
  - libstdcxx-ng=9.2.0=hdf63c60_2
  - ncurses=6.1=hf484d3e_1002
  - openssl=1.1.1d=h516909a_0
  - pip=20.0.2=py_2
  - pycosat=0.6.3=py37h516909a_1002
  - pycparser=2.19=py37_1
  - pyopenssl=19.1.0=py37_0
  - pysocks=1.7.1=py37_0
  - python=3.7.6=h357f687_2
  - readline=8.0=hf8c457e_0
  - requests=2.22.0=py37_1
  - ruamel_yaml=0.15.80=py37h516909a_1000
  - setuptools=45.2.0=py37_0
  - six=1.14.0=py37_0
  - sqlite=3.30.1=hcee41ef_0
  - tk=8.6.10=hed695b0_0
  - tqdm=4.42.1=py_0
  - urllib3=1.25.7=py37_0
  - wheel=0.34.2=py_1
  - xz=5.2.4=h14c3975_1001
  - yaml=0.2.2=h516909a_1
  - zlib=1.2.11=h516909a_1006
  - pip:
    - click==8.0.1
    - flask==2.0.1
    - importlib-metadata==4.5.0
    - itsdangerous==2.0.1
    - jinja2==3.0.1
    - markupsafe==2.0.1
    - typing-extensions==3.10.0.0
    - werkzeug==2.0.1
    - zipp==3.4.1
prefix: /opt/conda
```

Is there any guarantee that a `conda env export` will **always** result in the same build result? I guess the line has to be drawn somewhere... conda will never be fully reproducible and that is okay but I would be nice to be able to refer to an environment by it's pinned sha256 versions of all package. So I'd like to support this.

So one gotcha is that we'd need to remove the name and prefix I think. I know long tangent but @leej3 I like the idea!

**Comment by costrouc** [comment]
Closing as implemented in #90 in https://github.com/Quansight/conda-store/blob/b260386d154d9394485e12fdf3ed1f3cc72a47ee/conda-store-server/conda_store_server/orm.py#L98
